### PR TITLE
deprecation version boundary

### PR DESCRIPTION
**Pull request by holmanb**
## Additional Context

Add deprecation version knob for stable release distros. This allows distros to set an "initial release" version for the cloud-init release on a stable distro, and any newly added deprecation in future releases should log at an INFO level. 

with the following cloud-config:
```yaml
#cloud-config
grub-dpkg:
  enabled:
    false
users:
  - name: osadmin
    lock-passwd: false
    sudo: ["ALL=(ALL) NOPASSWD:ALL"]
    ssh-authorized-keys:
      - ssh-rsa "AAAAB3NzaC1yc2EAAAADAQABAAABAQCc/K1T62elOkJKc94sWovn06RE7D274Fxx5n+eT/oClAwk7632awMtWjO+lIRpzOlEcYCR6wexlmL9mZNEFJoROMMRyCxRilNAiDKOq0X5j1HXY2ky9KjtLxo2b8yXFRBcYKK4W8aHU94NkNek81jdi9lC+L6jUrZ25d7xylqnEbd4TgrBphiowdh5B3RY+j6ePthLWWhHopRcVDgq91yYacelCCIfSDaGBya9iXPZSwoKtPin4n5PdSGmSOA3fIkvV5JjLS1QbuY5tzhXKuH9MdAqrvutF1bprs/GDf0IY8DO7DDoo8rXjhhgGtCdg1i5UitZIAM3hn/Vk5F//5eJ JOOS8312@EB-OR6120158"

```
this PR yields the following schema annotate output:
```yaml
# cloud-init schema --system --annotate
Found cloud-config data types: user-data, network-config

1. user-data at /var/lib/cloud/instances/cb88d1af-f827-4381-bc89-4fad787bcaaf/cloud-config.txt:
#cloud-config

# from 1 files
# part-001

---
grub-dpkg:		# D1
    enabled: false
users:
-   lock-passwd: false		# D2
    name: osadmin
    ssh-authorized-keys:		# D3
    - ssh-rsa "AAAAB3NzaC1yc2EAAAADAQABAAABAQCc/K1T62elOkJKc94sWovn06RE7D274Fxx5n+eT/oClAwk7632awMtWjO+lIRpzOlEcYCR6wexlmL9mZNEFJoROMMRyCxRilNAiDKOq0X5j1HXY2ky9KjtLxo2b8yXFRBcYKK4W8aHU94NkNek81jdi9lC+L6jUrZ25d7xylqnEbd4TgrBphiowdh5B3RY+j6ePthLWWhHopRcVDgq91yYacelCCIfSDaGBya9iXPZSwoKtPin4n5PdSGmSOA3fIkvV5JjLS1QbuY5tzhXKuH9MdAqrvutF1bprs/GDf0IY8DO7DDoo8rXjhhgGtCdg1i5UitZIAM3hn/Vk5F//5eJ
        JOOS8312@EB-OR6120158"
    sudo:
    - ALL=(ALL) NOPASSWD:ALL
...

# Deprecations: -------------
# D1: An alias for ``grub_dpkg`` Deprecated in version 22.2. Use ``grub_dpkg`` instead.
# D2: Default: ``true`` Deprecated in version 22.3. Use ``lock_passwd`` instead.
# D3:  Deprecated in version 18.3. Use ``ssh_authorized_keys`` instead.


  Valid schema user-data

2. network-config at /var/lib/cloud/instances/cb88d1af-f827-4381-bc89-4fad787bcaaf/network-config.json:
  Valid schema network-config
```
and the deprecated key use is logged at level DEPRECATED:

```bash
# cloud-init status  --long
status: done
extended_status: degraded done
boot_status_code: enabled-by-generator
last_update: Thu, 01 Jan 1970 00:00:04 +0000
detail: DataSourceLXD
errors: []
recoverable_errors:
DEPRECATED:
	- Deprecated cloud-config provided:grub-dpkg: An alias for ``grub_dpkg`` Deprecated in version 22.2. Use ``grub_dpkg`` instead., users.0.lock-passwd: Default: ``true`` Deprecated in version 22.3. Use ``lock_passwd`` instead., users.0.ssh-authorized-keys:  Deprecated in version 18.3. Use ``ssh_authorized_keys`` instead.
```
when DEPRECATION_INFO_BOUNDARY is set to 22.2
```diff
diff --git a/cloudinit/features.py b/cloudinit/features.py
index 9d88e138c..7f6900969 100644
--- a/cloudinit/features.py
+++ b/cloudinit/features.py
@@ -87,7 +87,7 @@ On Debian and Ubuntu systems, cc_apt_configure will write a deb822 compatible
 to write /etc/apt/sources.list directly.
 """
 
-DEPRECATION_INFO_BOUNDARY = "devel"
+DEPRECATION_INFO_BOUNDARY = "22.2"
 """
 DEPRECATION_INFO_BOUNDARY is used by distros to configure at which version to
 start logging deprecations at a level higher than INFO.
```

all deprecations are still reported
```
root@me:~# cloud-init schema --annotate --system
Found cloud-config data types: user-data, network-config

1. user-data at /var/lib/cloud/instances/5362d6fe-c68b-4e53-9920-49795eac9c28/cloud-config.txt:
#cloud-config

# from 1 files
# part-001

---
grub-dpkg:		# D1
    enabled: false
users:
-   lock-passwd: false		# D2
    name: osadmin
    ssh-authorized-keys:		# D3
    - ssh-rsa "AAAAB3NzaC1yc2EAAAADAQABAAABAQCc/K1T62elOkJKc94sWovn06RE7D274Fxx5n+eT/oClAwk7632awMtWjO+lIRpzOlEcYCR6wexlmL9mZNEFJoROMMRyCxRilNAiDKOq0X5j1HXY2ky9KjtLxo2b8yXFRBcYKK4W8aHU94NkNek81jdi9lC+L6jUrZ25d7xylqnEbd4TgrBphiowdh5B3RY+j6ePthLWWhHopRcVDgq91yYacelCCIfSDaGBya9iXPZSwoKtPin4n5PdSGmSOA3fIkvV5JjLS1QbuY5tzhXKuH9MdAqrvutF1bprs/GDf0IY8DO7DDoo8rXjhhgGtCdg1i5UitZIAM3hn/Vk5F//5eJ
        JOOS8312@EB-OR6120158"
    sudo:
    - ALL=(ALL) NOPASSWD:ALL
...

# Deprecations: -------------
# D1: An alias for ``grub_dpkg`` Deprecated in version 22.2. Use ``grub_dpkg`` instead.
# D2: Default: ``true`` Deprecated in version 22.3. Use ``lock_passwd`` instead.
# D3:  Deprecated in version 18.3. Use ``ssh_authorized_keys`` instead.


  Valid schema user-data

2. network-config at /var/lib/cloud/instances/5362d6fe-c68b-4e53-9920-49795eac9c28/network-config.json:
  Valid schema network-config
```
however only older deprecations than this version are logged at a `DEPRECATED` level:
```
root@me:~# cloud-init status --long
status: done
extended_status: degraded done
boot_status_code: enabled-by-generator
last_update: Thu, 01 Jan 1970 00:00:03 +0000
detail: DataSourceLXD
errors: []
recoverable_errors:
DEPRECATED:
	- Deprecated cloud-config provided:grub-dpkg: An alias for ``grub_dpkg`` Deprecated in version 22.2. Use ``grub_dpkg`` instead.
```

nice to have:

- [ ] log the release boundary
- [ ] user configurable release boundary

## Checklist

- [ ] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/index.html)
- [ ] I have updated or added any [unit tests](https://cloudinit.readthedocs.io/en/latest/development/testing.html) accordingly
- [ ] I have updated or added any [documentation](https://cloudinit.readthedocs.io/en/latest/development/contribute_docs.html) accordingly

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

File: cloudinit/util.py [hunk 1]
```diff
@@ -3242,14 +3251,20 @@ def deprecate(
         f"{deprecated_version} and scheduled to be removed in "
         f"{version_removed}. {message}"
     ).rstrip()
-    if return_log:
-        return deprecate_msg
-    if dedup not in deprecate._log:  # type: ignore
+    if (
```
Comment (holmanb) [inline cloudinit/util.py:3254]: This logic isn't beautiful. With 3.10, I might reach for a `match` statement for more expressive logic, but that's a few years out.

Ideas welcome.

Comment (TheRealFalcon) [reply cloudinit/util.py:3254]: Looks fine to me.

**Review by TheRealFalcon — COMMENTED** [review]
Starting the bikeshed :wink: 

Looks good so far though

File: cloudinit/settings.py [hunk 2]
```diff
@@ -78,3 +78,27 @@
...
+# first version released in their stable distro. By doing this, they can expect
+# that newly added deprecations will be logged at INFO level.
+#
+# format:
+#
+# <value> :: = <default> | <version>
+# <default> ::= "devel"
+# <version> ::= <major> "." <minor> "." "patch"
+#
+DEPRECATION_INFO_BOUNDARY = "devel"
```
[the remark this answers is not in the document]
Comment (blackboxsw) [reply cloudinit/settings.py:104]: Good thought on features.py.

File: cloudinit/settings.py [hunk 3]
```diff
@@ -78,3 +78,27 @@
...
+# cloud-init will use that version as a demarcation point. Deprecations which
+# are added after this version will be logged as at an INFO level. Deprecations
+# which predate this version will be logged at the higher DEPRECATED level.
+# Downstreams that want stable log behavior may set the variable below to the
+# first version released in their stable distro. By doing this, they can expect
+# that newly added deprecations will be logged at INFO level.
+#
+# format:
+#
+# <value> :: = <default> | <version>
```
Comment (blackboxsw) [inline cloudinit/settings.py:100]:
```suggestion
# <value> :: = devel | <version>
```

Comment (blackboxsw) [reply cloudinit/settings.py:100]: I guess you were expressing that default == "devel" on the next line. so no need to accept this comment

Comment (holmanb) [reply cloudinit/settings.py:100]: correct, we probably don't need to express this in BNF, but it's what came to my mind in the moment.

Now that I look though the patch part is wrong, I'll fix that.

**Review by blackboxsw — CHANGES_REQUESTED** [review]
Looks good some unittest fixes to mock the feature DEPRECATION_INFO_BOUNDARY but otherwise +1

- Would like to see a unit test that exercise a DEPRECATION_INFO_BOUNDARY of something != "devel" to provide examples to other downstream of what version strings appear acceptable and the behavior expected in those cases.

File: cloudinit/config/schema.py [hunk 4]
```diff
@@ -785,25 +794,39 @@ def validate_cloudconfig_schema(
...
             schema_error, SchemaDeprecationError
         ):  # pylint: disable=W1116
-            deprecations += problem
+            if (
+                "devel" != features.DEPRECATION_INFO_BOUNDARY
+                and Version.from_str(schema_error.version) >
+                Version.from_str(features.DEPRECATION_INFO_BOUNDARY)
+            ):
+                info_deprecations.append(
+                    SchemaProblem(path, schema_error.message)
```
[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/config/schema.py:806]: The original code created a tuple with a single element and then appended this single element to a list:

```python
problem = (SchemaProblem(path, schema_error.message),)
...
deprecations += problem
```

which takes some extra mental gymnastics to realize the end goal is to append to a list

```python
deprecations.append(SchemaProblem(path, schema_error.message))
```

I consider the legibility difference between:

```python
if <blah>:
    info_deprecations.append(SchemaProblem(path, schema_error.message))
else:
    deprecations.append(SchemaProblem(path, schema_error.message))
```

and

```python
problem = SchemaProblem(path, schema_error.message)
if <blah>:
    info_deprecations.append(problem)
else:
    deprecations.append(problem)
```
to be negligible. The second has less total calls to grok, but the first has an extra variable to track, and since the calls are simple and identical it's quick realize that they do the same thing.

> what's the perspective on dropping the local var declaration 'problem =' 

It's really just a style preference. When I find differences like this negligible I typically lean towards the option that involves assigning fewer variables for a few of reasons:

1) less state / context to track -> easier to mentally follow
2) easier to refactor into simpler / cleaner code in the future
3) less variable names that I need to think up (obviously doesn't apply in this case because pre-existing code)
4) less lines of code to read / maintain

While these reasons reasons seem bit trivial, the impacts add up as code size and complexity grows.

I suppose I could have done this PR without making this change, but the actual reason that this happened due to an earlier iteration subclassing a different type and having to move the assignment within the `if isinstance()` call which necessitated modifying this variable anyways.

File: cloudinit/features.py [hunk 5]
```diff
@@ -87,6 +87,31 @@
...
+which predate this version will be logged at the higher DEPRECATED level.
+Downstreams that want stable log behavior may set the variable below to the
+first version released in their stable distro. By doing this, they can expect
+that newly added deprecations will be logged at INFO level.
+
+format:
+
+<value> :: = <default> | <version>
+<default> ::= "devel"
+<version> ::= <major> "." <minor> "." <patch>
```
[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/features.py:112]: > Now that we are accepting a slightly more complex schema than booleans in features.py, I bet if we get more complex values for other 'features', some sort of schema validation of the configured values could be helpful to avoid the potential of downstreams. It's definitely not worth the investment on this PR, but a discussion topic we could bring up if we get more complex configurable feature values.

Eh, maybe. Personally I think that trying to prevent errors introduced by downstreams or third-parties in upstream code is a lot lower priority than [validating user-introduced errors and documenting the possible cloud.cfg values](https://github.com/canonical/cloud-init/issues/5410).

File: cloudinit/features.py [hunk 6]
```diff
@@ -87,6 +87,31 @@
 to write /etc/apt/sources.list directly.
 """
 
+DEPRECATION_INFO_BOUNDARY = "devel"
+"""
+DEPRECATION_INFO_BOUNDARY is used by distros to configure at which version to
+start logging deprecations at a level higher than INFO.
+
+The default value "devel" tells cloud-init to log all deprecations higher
+than INFO. This value may be overriden by downstreams in order to maintain
```
[the remark this answers is not in the document]
Comment (holmanb) [reply cloudinit/features.py:96]: I don't see the difference between the suggestion and current values?

File: cloudinit/config/schema.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (blackboxsw) [reply cloudinit/config/schema.py:806]: Good intent/reasoning here thanks.

**Comment by holmanb** [comment]
@blackboxsw I just need to add a unittest for this behavior and then I think it's ready

**Comment by blackboxsw** [comment]
> @blackboxsw I just need to add a unittest for this new behavior and then I think it's ready

Since I added the request for more testing, I figured I'd setup a couple of minor tests to exercise the boundary to get better context on behavior here. Review away, if you like the tests, let's keep em, or of something needs to change go for it.

**Review by blackboxsw — APPROVED** [review]

File: tests/unittests/test_log.py [hunk 7]
```diff
@@ -1,12 +1,13 @@
 # This file is part of cloud-init. See LICENSE file for license information.
 
 """Tests for cloudinit.log"""
-
```
Comment (holmanb) [inline tests/unittests/test_log.py:4]: unnecessary whitespace change
